Add login page

tamirmatok opened the pull request:
### Desired Outcome
Adding a landing page that is a login page to the Hotails website.
In order for a user to be able to access the site (whether it is a dog owner or a daycare owner) he must be a user and enter his details on this page.
Add login and logout funcenallity.

### Implemented Changes
Change Landing page to login page.
After sucsessful log-in the application recognize the user with all of his details.

### Connected Issue/Story

closes #57 #61 

![login_page](https://user-images.githubusercontent.com/95190977/163681465-39fbab3b-0d44-4aee-bf2a-750f4a2c8850.JPG)

ErezCohenn reviewed (changes requested):
Great Job @tamirmatok !!!
few things: 
- I saw u deleted the test `test_root_entrypoint`. In that, this PR should close
 #61 
and I think that it needed to be in another commit (another option is that I will open a new PR for that issue because I was assigned for that)
- consider uploading some screenshots for the new UI login
- In the PR description you wrote: "We want that **our** app landing page will be a login page"
 but the content should not have attributions. I think you should write something like this:
"Adding a landing page that is a login page to the Hotails website.
In order for a user to be able to access the site (whether it is a dog owner or a daycare owner) he must be a user owner and enter his details on this page.

OfirMatasas commented on Hotails/urls.py:
```diff
@@ -14,10 +14,12 @@
+
     path('admin/', admin.site.urls),
-    path('', views.homepage, name='homepage')
+    path('', include('main.urls')),
```
Django will fire the first view for which the URL matches. It thus evaluates the URL top-to-bottom.
So when you have same 2 URL patterns, the bottom one is useless.

I suggest you switch them over and include the "homepage" in main.urls.

OfirMatasas commented on Pipfile:
```diff
@@ -5,6 +5,7 @@ name = "pypi"
 
 [packages]
 django = "*"
+django-crispy-forms = "*"
```
What is this crispy -forms?

OfirMatasas commented on main/templates/main/login.html:
```diff
@@ -0,0 +1,49 @@
+            {{message}} <button type="button" class="btn-close" data-bs-dismiss="alert" aria-label="Close"></button>
+        </div>
+        {% endfor %}
+        {% endif %}
```
Why do we need this check in the login page?

tamirmatok replied:
It is useful when you need to display a one-time notification message to the user after processing a form or some other types of user input( 'django.contrib.messages' is in [INSTALLED_APPS]).
For now, to make thing simple , I removed it .

tamirmatok commented:
> Great Job @tamirmatok !!! few things:
> 
> * I saw u deleted the test `test_root_entrypoint`. In that case, this PR should close #61
>   and I think that it needed to be in another commit (another option is that I will open a new PR for that issue because I was assigned for that)
> * consider uploading some screenshots for the new UI login
> * In the PR description you wrote: "We want that **our** app landing page will be a login page"
>   but the content should not have attributions. I think you should write something like this:
>   "Adding a landing page that is a login page to the Hotails website.
>   In order for a user to be able to access the site (whether it is a dog owner or a daycare owner) he must be a user owner and enter his details on this page.

Hi @ErezCohenn ,
1) I have deleted this file because of 2 resasons
- I saw #61 
- After my changes he failed in test - the reason for it beacuse we got 302 response code instead of 200 (because we are redirecting to login page) 
So I think we can leave it like that and I can seperate it to another commit and close #61 , unless you think it deserve seperate PR (in that case I will put this test in comment instead of deletion).
2) Will be send in Slack.
3)Thanks, just fixed .

ErezCohenn commented:
> > Great Job @tamirmatok !!! few things:
> > 
> > * I saw u deleted the test `test_root_entrypoint`. In that case, this PR should close [Entrypoint test does not test what it needs too #61](https://github.com/redhat-beyond/Hotails/issues/61)
> >   and I think that it needed to be in another commit (another option is that I will open a new PR for that issue because I was assigned for that)
> > * consider uploading some screenshots for the new UI login
> > * In the PR description you wrote: "We want that **our** app landing page will be a login page"
> >   but the content should not have attributions. I think you should write something like this:
> >   "Adding a landing page that is a login page to the Hotails website.
> >   In order for a user to be able to access the site (whether it is a dog owner or a daycare owner) he must be a user owner and enter his details on this page.
> 
> Hi @ErezCohenn ,
> 
> 1. I have deleted this file because of 2 resasons
> 
> * I saw [Entrypoint test does not test what it needs too #61](https://github.com/redhat-beyond/Hotails/issues/61)
> * After my changes he failed in test - the reason for it beacuse we got 302 response code instead of 200 (because we are redirecting to login page)
>   So I think we can leave it like that and I can seperate it to another commit and close [Entrypoint test does not test what it needs too #61](https://github.com/redhat-beyond/Hotails/issues/61) , unless you think it deserve
>   seperate PR (in that case I will put this test in comment instead of deletion).
> 
> 2. Will be send in Slack.
> 3. Thanks, just fixed .

1. I think it's ok to leave it like its right now and remain the deletion of the entry_point test in this PR, but I think you should split it to another commit.
and don't forget to mention it in the description of the PR that this PR is closing also #61 :)

tamirmatok commented:
> > > Great Job @tamirmatok !!! few things:
> > > 
> > > * I saw u deleted the test `test_root_entrypoint`. In that case, this PR should close [Entrypoint test does not test what it needs too #61](https://github.com/redhat-beyond/Hotails/issues/61)
> > >   and I think that it needed to be in another commit (another option is that I will open a new PR for that issue because I was assigned for that)
> > > * consider uploading some screenshots for the new UI login
> > > * In the PR description you wrote: "We want that **our** app landing page will be a login page"
> > >   but the content should not have attributions. I think you should write something like this:
> > >   "Adding a landing page that is a login page to the Hotails website.
> > >   In order for a user to be able to access the site (whether it is a dog owner or a daycare owner) he must be a user owner and enter his details on this page.
> > 
> > 
> > Hi @ErezCohenn ,
> > 
> > 1. I have deleted this file because of 2 resasons
> > 
> > 
> > * I saw [Entrypoint test does not test what it needs too #61](https://github.com/redhat-beyond/Hotails/issues/61)
> > * After my changes he failed in test - the reason for it beacuse we got 302 response code instead of 200 (because we are redirecting to login page)
> >   So I think we can leave it like that and I can seperate it to another commit and close [Entrypoint test does not test what it needs too #61](https://github.com/redhat-beyond/Hotails/issues/61) , unless you think it deserve
> >   seperate PR (in that case I will put this test in comment instead of deletion).
> > 
> > 
> > 2. Will be send in Slack.
> > 3. Thanks, just fixed .
> 
> 1. I think it's ok to leave it like its right now and remain the deletion of the entry_point test in this PR, but I think you should split it to another commit.
>    and don't forget to mention it in the description of the PR that this PR is closing also [Entrypoint test does not test what it needs too #61](https://github.com/redhat-beyond/Hotails/issues/61) :)

In second thought, Im not sure it deserve seperate commit , I have added suitable description in PR description. @ErezCohenn

OfirMatasas reviewed (approved):
LGTM

ErezCohenn reviewed (changes requested):
I think you should change the name of the files: `fffff-min.png` and nicepage.css` to a meaningful name

Omeramsc commented on main/templates/main/login.html:
```diff
@@ -0,0 +1,41 @@
+    <title>Hotails.com - Login!</title>
+</head>
+
+<body>
```
Do note this will need to be changed once the navbar is implemented

tamirmatok replied:
Hi @Omeramsc , why is that ? 
we don’t want that navbar will appear at login page .

Omeramsc replied:
The Navbar should appear on all pages as it is a part of the base template. you should include logic to hide the buttons when the user is not logged in (or not of the right type)
there are exceptions to the rule, but I highly suggest you'll implement it like this on this project.

Omeramsc commented on main/tests.py:
```diff
@@ -1 +1,62 @@
+@pytest.mark.django_db
+class TestLoginView:
+
+    def test_enter_login_page(self, client):
```
note #61
The logic in the page should be tested, not the status code

Omeramsc replied:
1. You can modify it so if the current user isn't logged in, it'll redirect from the entrypoint ("/") to the login page. *this* logic deserves a test.
2. This test doesn't really test the HTTP connectivity as it uses client, which calls Django WSGI layers directly. your tests need to check what is shown on the page itself.

Omeramsc commented on main/tests.py:
```diff
@@ -1 +1,62 @@
+        response = client.get('/login/')
+        assert response.status_code == 200
+        logged_user = response.wsgi_request.user
+        assert logged_user == dog_owner_user.user
```
also save a logged_used before force_login and test that they are not the same.

tamirmatok replied:
Done.

Omeramsc commented on main/tests.py:
```diff
@@ -1 +1,62 @@
+                                         'user', 1234567890, 'dog race', 'https://www.google.com/user1', 4, 2, 'M')
+
+        client.force_login(dog_owner_user.user)
+        response = client.get('/login/')
```
what is this needed for?

tamirmatok replied:
Deleted .

Omeramsc commented on main/tests.py:
```diff
@@ -1 +1,62 @@
+        response = client.get('/login/')
+        assert response.status_code == 200
+        logged_user = response.wsgi_request.user
+        assert logged_user == day_care_user.user
```
apply comments of former test

Omeramsc commented on main/tests.py:
```diff
@@ -1 +1,62 @@
+                }
+        response = client.post('/login/', form, follow=True)
+        assert response.status_code == 200
+        assert dog_owner_user.user == response.wsgi_request.user
```
also save a logged_used before force_login and test that they are not the same.

tamirmatok replied:
Fixed.

Omeramsc commented on main/tests.py:
```diff
@@ -1 +1,62 @@
+        form = {'username': 'dogOwnerUser02',
+                'password': 'password123',
+                }
+        response = client.post('/login/', form, follow=True)
```
I'm not sure the former test is needed, this covers it, doesn't it?

tamirmatok replied:
You right , it seems that client response after force_login method is indeed include in this test.

Omeramsc commented on main/tests.py:
```diff
@@ -1 +1,62 @@
+        assert logged_user == day_care_user.user
+
+    def test_valid_login_dog_owner_user_info(self, client):
+        dog_owner_user = DogOwner.create('dogowner@address.com', 'dogOwnerUser02', 'password123', 'dog name', 'test',
```
may want to fixture this

Omeramsc commented on main/tests.py:
```diff
@@ -1 +1,62 @@
+        assert dog_owner_user.user == response.wsgi_request.user
+
+    def test_invalid_login_day_care_user_info(self, client):
+        day_care_user = DayCare.create('daycare@address.com', 'dayCareUser', 'password123', 'daycare',
```
may want to fixture this

Omeramsc commented on main/views.py:
```diff
@@ -1,5 +1,26 @@
+# from dogowner.models import DogOwner
+# from daycare.models import DayCare
+# from daycare.views import daycare_home
+# from dogowner.views import dog_owner_home
```
remove unused imports

Omeramsc commented on main/views.py:
```diff
@@ -1,5 +1,26 @@
-def homepage(request):
-    return render(request, 'main/homepage.html')
+def index(request):
+    # need to implement home function for logged- in user.
```
seems like this PR is not ready for review yet.
please either:
1.  implement, including tests
2.  redirect to a different, logical place, then open an issue to implement this function instead of comments in the code.

Omeramsc commented on main/views.py:
```diff
@@ -1,5 +1,26 @@
+    # put day-care home page view function here.
+    # else DayCare.objects.filter(user=request.user).exists():
+    #     return daycare_home(request)
+    pass
```
same

Omeramsc commented:
also: don't forget to fix the conflicts.

OfirMatasas commented on main/static/css/style.css:
```diff
@@ -0,0 +1,94 @@
+body {
```
The name of the file is not good.
The name of the CSS file should be the same as the HTML and in the same directory with it, or in a directory for stylesheets, for better orientation.

tamirmatok replied:
Fixed.

jeniaSakirko reviewed (approved):
LGTM

Omeramsc commented on main/templates/main/login.html:
```diff
@@ -0,0 +1,38 @@
+{% endblock %}
+
+{% block content %}
+
```
You may also want to block this page from a logged-in user view (either hide the content or find a way to make it inaccessible)

tamirmatok replied:
Fixed by adding "redirect_authenticated_user=True" in Login view .

Omeramsc commented on main/tests.py:
```diff
@@ -1 +1,73 @@
+        assert response.status_code == 302
+        assert response['Location'] == '/homepage/'
+
+# # Create your tests here.
```
can remove this comment

tamirmatok replied:
Done.

Omeramsc commented on main/tests.py:
```diff
@@ -1 +1,73 @@
+
+    def test_root_entrypoint_redirection_logged_user(self, client, create_dog_owner_user):
+        client.force_login(user=create_dog_owner_user.user)
+        response = client.get("/")
```
how about testing that a logged user trying to reach /login/ will reach the homepage instead?

tamirmatok replied:
Done , I added test in TestLoginView that test this case but redirect to the entry point (And from there we have test that cover redirect of logged-in user to homepage ) , because in the future we will have 2 home pages (DogOwner/DayCare) and it is looks good to me to manage this case from index view.

Omeramsc replied:
I'm not sure "2 home pages" is the right approach or phrasing.

tamirmatok replied:
You right . I meant 2 views for diffrent kind of users.

Omeramsc commented on main/tests.py:
```diff
@@ -1 +1,73 @@
+class TestIndexView:
+
+    def test_root_entrypoint_redirection_unlogged_user(self, client):
+        response = client.get("/")
```
what about pages that are not the entrypoint? say the user type the URL to access it directly?

worth a test IMO.

tamirmatok replied:
For now we have only homepage and login pages , and homepage is under @login_required decorator,
do you think I should test this funconallity ?

Omeramsc commented on main/views.py:
```diff
@@ -1,5 +1,21 @@
+def index(request):
+    if request.user.is_authenticated:
+        return redirect(to='homepage')
+    else:
```
else is redundant.

tamirmatok replied:
Fixed.

tamirmatok commented on main/templates/main/base_template.html:
```diff
@@ -21,15 +24,19 @@
+                    <li class="nav-item"><a class="nav-link text-white" href="">Orders</a></li>
+                    <li class="nav-item"><a class="nav-link text-white" href="">Search</a></li>
+                    <li class="nav-item"><a class="nav-link text-white" href="">Chats</a></li>
+                {% endif %}
```
It seems that the solution for that is to add for every future view @login_required decorator.  @Omeramsc

Omeramsc commented on static/css/Home.css:
```diff
@@ -267,7 +267,7 @@
 .u-section-2 .u-image-1 {
   min-height: 655px;
-  background-image: url("images/fffff-min.jpg");
+  background-image: url("../images/fffff-min.jpg");
```
please change the name of this photo

tamirmatok replied:
Done .

tamirmatok replied:
Done.

Omeramsc reviewed (approved):
LGTM

Yuval-Vino reviewed (approved):
Great job Tamir, thank you for taking care of this!

ormergi commented on main/tests.py:
```diff
@@ -1 +1,84 @@
+                 }
+        response = client.post('/login/', forms, follow=True)
+        assert create_dog_owner_user.user != response.wsgi_request.user
+        assert b"Please enter a correct username and password." in response.content
```
Asserting error messages is not a good practice and should be avoided when possible,
the reason for that is the fact that if the component that raises it changes you also need to change your tests.

Is there a way to assert that there was an error w/o asserting the error message?

tamirmatok replied:
I think in that case just to check the response.wsgi_request.user user is not the user given in the form .

ormergi commented on main/tests.py:
```diff
@@ -1 +1,84 @@
+
+@pytest.mark.django_db
+class TestIndexView:
+
```
nit: redundant newline

ormergi commented on main/tests.py:
```diff
@@ -1 +1,84 @@
+
+
+class TestHomepageView:
+
```
nit: redundant newline

ormergi commented on main/urls.py:
```diff
@@ -0,0 +1,12 @@
+from django.contrib.auth import views as auth_views
+from . import views
+
+urlpatterns = [
```
I would move this list to be used directly at `Hotails/url.py` to improve visibility (so that you wont need to just from one file to another to understand the url and routs)

tamirmatok replied:
Done.

ormergi commented on tests/test_root_entrypoint.py:
```diff
@@ -1,3 +0,0 @@
-def test_root_entrypoint(client):
```
I think this will close https://github.com/redhat-beyond/Hotails/issues/61

tamirmatok replied:
Thats right , I have mentioned it in PR "Connected Issue/Story" section .

ormergi commented on main/tests.py:
```diff
@@ -1 +1,84 @@
+
+@pytest.mark.django_db
+class TestLoginView:
+
```
nit: redundant newline

ormergi commented on main/tests.py:
```diff
@@ -1 +1,84 @@
+                }
+
+        response = client.post('/login/', form, follow=True)
+        curr_log_user = response.wsgi_request.user
```
Please avoid shortening variable names when possible.

ormergi reviewed (changes requested):
Hi Tamir, 
Please avoid describing implementation details at the PR description and commit message (e.g modify settings.py, cover with tests, etc..) it should be understood from the diff (actual PR changes).
Also please change the commit message body to describe what is going to be changed (similar to the into at the PR description) and why (when it make sense).

ormergi reviewed (approved):
Thanks for the changes, LGTM
